Re-implement sidecar module as http_gateway

reset opened the pull request:
* Remove sidecar module and replace with http_gateway
* Refactorings to allow state of Manager to be shared with http gateway
* Add initial raml documentation for supervisor http gateway

![gif-keyboard-7839051923796985666](https://cloud.githubusercontent.com/assets/54036/20596756/ceea2d3a-b1f4-11e6-9325-1da4831ef4f4.gif)

thesentinels commented:
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

adamhjk commented on components/butterfly/src/message/mod.rs:
```diff
@@ -41,3 +44,117 @@ pub fn unwrap_wire(payload: &[u8], ring_key: &Option<SymKey>) -> Result<Vec<u8>>
     }
 }
+
+impl Encodable for swim::Election {
```
I love how this comes together.

adamhjk commented on components/sup/doc/api.raml:
```diff
@@ -0,0 +1,75 @@
+---
+title: Habitat Supervisor API
+
+baseUri: https://{rootUri}
```
This should be http, not https, unless we're auto-generating certificates. That's probably a neat idea, actually, but it wouldn't prove a ton security wise.

adamhjk replied:
Fixed

adamhjk commented on components/sup/src/http_gateway.rs:
```diff
@@ -0,0 +1,194 @@
+use manager::{self, census};
+use util::signals;
+
+static LOGKEY: &'static str = "HG";
```
![gif-keyboard-12254865723829879485](https://cloud.githubusercontent.com/assets/4304/20642719/57e93b08-b3cb-11e6-8023-b34d0172d5ac.gif)

adamhjk commented on components/sup/src/http_gateway.rs:
```diff
@@ -0,0 +1,194 @@
+
+impl Default for ListenAddr {
+    fn default() -> ListenAddr {
+        ListenAddr(SocketAddrV4::new(Ipv4Addr::new(0, 0, 0, 0), 9631))
```
Should this be configurable, or a constant?

reset replied:
This will inline so we don't need to move it to a constant. This is literally the "default ip addr/port" for the http_gateway so I don't think this should be configurable here, instead we configure by the CLI & assign it to the Config struct. This is just what will be populated into that config struct by default

adamhjk commented on components/sup/src/http_gateway.rs:
```diff
@@ -0,0 +1,194 @@
+impl FromStr for ListenAddr {
+    type Err = SupError;
+
+    fn from_str(val: &str) -> Result<Self> {
```
Why not just use the normal SocketAddr traits, and let this work for both v4 and v6?

Specifically: https://doc.rust-lang.org/std/net/enum.SocketAddr.html should let you take an arbitrary v4 or v6 socket.

reset replied:
Great suggestion - I totally didn't know about that enum!

adamhjk commented on components/sup/src/http_gateway.rs:
```diff
@@ -0,0 +1,194 @@
+}
+
+fn butterfly(req: &mut Request) -> IronResult<Response> {
+    let state = req.get::<persistent::Read<ManagerState>>().unwrap();
```
turbofish to the rescue.

adamhjk commented on components/sup/src/http_gateway.rs:
```diff
@@ -0,0 +1,194 @@
+fn config(req: &mut Request) -> IronResult<Response> {
+    let state = req.get::<persistent::Read<ManagerState>>().unwrap();
+    let service_group =
+        ServiceGroup::new(req.extensions.get::<Router>().unwrap().find("svc").unwrap(),
```
What happens when this panics? Can it essentially not panic because of the shape of the route?

reset replied:
Exactly, you can't panic here - the handler would never be hit

adamhjk commented on components/sup/src/main.rs:
```diff
@@ -153,31 +152,16 @@ fn config_from_args(subcommand: &str, sub_args: &ArgMatches) -> Result<()> {
-    config.set_http_listen_ip(sidecar_ip);
-    config.set_http_listen_port(sidecar_port);
-
+    if let Some(addr_str) = sub_args.value_of("listen-http") {
```
yeah - since this is stringy, I think my comment above about just taking `SocketAddr` rather than a specific v4 is the right thing, as we will then work transparently with v6.

adamhjk commented on components/sup/src/manager/mod.rs:
```diff
@@ -15,35 +15,54 @@
 
-#[derive(Debug)]
+#[derive(Clone)]
+pub struct State {
```
This turned out very nicely.

adamhjk commented on components/sup/src/manager/mod.rs:
```diff
@@ -183,8 +193,8 @@ impl Manager {
     pub fn check_for_updated_packages(&mut self) {
-        for service in self.services.iter_mut() {
-            service.check_for_updated_package();
+        for service in self.state.services.write().unwrap().iter_mut() {
```
Man, I hate taking this lock so frequently for an operation that almost never will actually write. Given how the code is written, I think it won't matter much, since we don't have a ton of parallel read locks outstanding, so the cost is probably minimal. But I pour one out for the lock-free services list. :)

adamhjk replied:
Should be expect

adamhjk commented on components/sup/src/manager/mod.rs:
```diff
@@ -104,74 +115,73 @@ impl Manager {
         match signals::check_for_signal() {
             Some(SignalEvent::Shutdown) => {
-                for service in self.services.iter_mut() {
+                for service in self.state.services.write().unwrap().iter_mut() {
```
These unwrap's all need to become `expect("Services lock is poisoned")` or something similar, rather than just the raw explosion. Makes a big difference in debug hunting - the raw unwrap tells you almost nothing about the lock you're digging for.

adamhjk commented on components/sup/src/manager/mod.rs:
```diff
@@ -104,74 +115,73 @@ impl Manager {
             }
             Some(SignalEvent::Passthrough(signal_code)) => {
-                for service in self.services.iter() {
+                for service in self.state.services.read().unwrap().iter() {
```
Should be `expect`

adamhjk commented on components/sup/src/manager/mod.rs:
```diff
@@ -240,22 +257,24 @@ impl Manager {
             }
 
-            for service in self.services.iter_mut() {
+            for service in self.state.services.write().unwrap().iter_mut() {
```
Should be `expect`

adamhjk commented on components/sup/src/manager/mod.rs:
```diff
@@ -240,22 +257,24 @@ impl Manager {
                 // Reconfigure if neccessary
                 if census_updated || service_config_updated {
-                    service.reconfigure(&self.census_list);
+                    service.reconfigure(&self.state.census_list.read().unwrap());
```
Should be `expect`.

adamhjk commented on components/sup/src/manager/mod.rs:
```diff
@@ -266,7 +285,7 @@ impl Manager {
                 // Start or restart the service
                 if service.needs_restart || service.is_down() {
-                    match service.restart(&self.census_list) {
+                    match service.restart(&self.state.census_list.read().unwrap()) {
```
Should be `expect`

adamhjk commented on components/sup/src/manager/service_updater.rs:
```diff
@@ -0,0 +1,143 @@
+    }
+}
+
+struct Worker {
```
This whole refactor is great.

adamhjk commented:
I'm going to carry this patch for you, @reset. I need your refactor of the manager and friends to land to avoid further merge conflicts as I add back in encryption to the service configuration and service files.

reset commented:
@adamhjk great code review, I didn't even know about the SocketAddr enum, thank you!

I addressed all of your comments. I totally didn't see your last message about handling this for me, sorry about that. I needed some programming time on this vacation anyway ;)

adamhjk commented:
@thesentinels merge

adamhjk commented:
@thesentinels approve

thesentinels commented:
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

thesentinels commented:
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/179099564).

adamhjk commented:
![gif-keyboard-2766689288864813007](https://cloud.githubusercontent.com/assets/4304/20643818/3b5d7ba0-b3d7-11e6-8e27-9c7e79803d19.gif)

thesentinels commented:
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/179099564).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

![gif-keyboard-3280869874741411265](https://cloud.githubusercontent.com/assets/4304/17755674/1e577b82-6490-11e6-96b7-1663d283c824.gif)
